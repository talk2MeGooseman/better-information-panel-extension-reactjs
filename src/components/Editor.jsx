import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import * as Showdown from "showdown";
import { SHOWDOWN_CONFIG } from "../services/constants";

import SimpleMDE from "react-simplemde-editor";
import "easymde/dist/easymde.min.css";

const TOOLBAR_FEATURES = [
  "heading-1",
  "heading-2",
  "heading-3",
  "bold",
  "italic",
  "strikethrough",
  "code",
  "quote",
  "unordered-list",
  "ordered-list",
  "table",
  "horizontal-rule",
  "guide",
  "preview"
];

export default class Editor extends React.Component {
  static propTypes = {
    tab: PropTypes.object.isRequired,
    tabsStore: PropTypes.object.isRequired,
    allowPreview: PropTypes.bool,
  };

  componentDidMount() {
    this.handleChange(this.props.tab.body);
    this.converter = new Showdown.Converter(SHOWDOWN_CONFIG);
  }

  handleChange = value => {
    if(this.props.tab.body !== value) {
      this.props.tabsStore.saveState = "";
    }
    this.props.tab.body = value;
  };

  render() {
    const { tab } = this.props;

    let toolbar = [...TOOLBAR_FEATURES];

    let Wrapper = styled.div`
      .editor-preview {
        background: ${tab.bgColor};
        color: ${tab.textColor};
      }
    `;

    if (this.props.allowPreview) {
      toolbar.push('preview');
    }

    return (
      <Wrapper>
        <SimpleMDE
          onChange={this.handleChange}
          value={this.props.tab.body}
          options={{
            autoDownloadFontAwesome: false,
            spellChecker: false,
            status: false,
            insertTexts: {
              toggleUnorderedList: ["* ", ""],
            },
            toolbar: toolbar,
            previewRender: (plainText) => {
              return this.converter.makeHtml(plainText); // Returns HTML from a custom parser
            },
          }}
        />
      </Wrapper>
    );
  }
}
