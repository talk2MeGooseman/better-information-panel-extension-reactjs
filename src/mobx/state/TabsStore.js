import { observable, computed, action, makeObservable } from "mobx";
import TabModel from "../model/TabModel";
import { uuid } from "../../services/Utils";
import {
  queryPanelInformation
} from "../../services/Ebs";
import {
  DEFAULT_BODY_TEXT,
  ACTIVE_STEP_1,
  ACTIVE_STEP_2,
  ACTIVE_STEP_3,
  ACTIVE_STEP_4,
} from "../../services/constants";
import { path } from 'ramda';

export default class TabsStore {
  tabs = [];
  client = null;
  saveState = "";
  loadingState = "pending";
  activeStep = ACTIVE_STEP_1;
  videoOverlayHeight = '80vh';
  videoComponentVisibility = true;
  videoComponentTransparent = true;
  videoToggleImageUrl = '';
  videoToggleButtonPosition = '';
  tabIndex = 0;

  constructor() {
    makeObservable(this, {
      tabs: observable,
      client: observable,
      saveState: observable,
      loadingState: observable,
      activeStep: observable,
      videoOverlayHeight: observable,
      videoComponentVisibility: observable,
      videoComponentTransparent: observable,
      videoToggleImageUrl: observable,
      videoToggleButtonPosition: observable,
      tabIndex: observable,
      tabCount: computed,
      fetchTabs: action,
      updateTabs: action,
      saveTabs: action
    });
  }

  get tabCount() {
    return this.tabs.length;
  }

  setVideoComponentVisability(value) {
    this.saveState = '';
    this.videoComponentVisibility = value;
  }

  setVideoComponentTransparency(value) {
    this.saveState = '';
    this.videoComponentTransparent = value;
  }

  setVideoToggleImageUrl(value) {
    this.saveState = '';
    this.videoToggleImageUrl = value;
  }

  setVideoToggleButtonPosition(value) {
    this.saveState = '';
    this.videoToggleButtonPosition = value;
  }

  addTab() {
    this.saveState = '';

    if (this.activeStep === ACTIVE_STEP_1) {
      this.activeStep = ACTIVE_STEP_2;
    }

    this.tabs.push(TabModel.fromJS(this, {
      id: uuid(),
      title: 'Add Title',
      textColor: '#000000',
      bgColor: '#ffffff',
      body: DEFAULT_BODY_TEXT,
    }));
  }

  fetchTabs() {
    this.loadingState = "pending"
    queryPanelInformation(this.client).then(
      // inline created action
      action("fetchSuccess", ({ data }) => {
        this.loadingState = "done"

        const result = path(['channel', 'bipConfig'], data)
        console.log(result)

        if (!result.tabs) {
          this.addTab();
          return;
        }

        this.activeStep = ACTIVE_STEP_4;

        // Check if the setting are defined from server, could be undefined
        // And screw shit up, use default state
        if (result.transparent !== undefined) {
          this.videoComponentTransparent = result.transparent;
        }
        if (result.visibility !== undefined) {
          this.videoComponentVisibility = result.visibility;
        }
        if (result.togglePosition !== undefined) {
          this.videoToggleButtonPosition = result.togglePosition;
        }

        this.tabs = result.tabs.map((tab) => {
          return TabModel.fromJS(this, tab);
        })
      }),
      // inline created action
      action("fetchError", error => {
        this.loadingState = "error"
      })
    )
  }

  updateTabs() {
    queryPanelInformation(this.client).then(
      // inline created action
      action("fetchSuccess", result => {
        if (!result.tabs) {
          this.addTab();
          return;
        }
        this.activeStep = ACTIVE_STEP_4;
        this.tabs = result.tabs.map((tab) => {
          return TabModel.fromJS(this, tab);
        })
      }),
      // inline created action
      action("fetchError", error => {
        this.loadingState = "error"
      })
    )
  }

  saveTabs() {
    this.saveState = "pending"
      setPanelInformation(this.token, this.toJSON()).then(
        // inline created action
        action("fetchSuccess", result => {
          this.saveState = "done"
          this.activeStep = ACTIVE_STEP_4;
        }),
        // inline created action
        action("fetchError", error => {
          this.saveState = "error"
        })
      )
  }

  toJSON() {
    let jTabs = this.tabs.map((tab) => {
      return tab.toJSON();
    });

    return {
      tabs: jTabs,
      visibility: this.videoComponentVisibility,
      transparent: this.videoComponentTransparent,
      togglePosition: this.videoToggleButtonPosition,
    };
  }
}
