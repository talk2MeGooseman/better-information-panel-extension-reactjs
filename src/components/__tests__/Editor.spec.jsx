import React from "react";
import { createShallow, createMount} from '@material-ui/core/test-utils';


import Editor from "../Editor";
import TabsStore from "../../mobx/state/TabsStore";

describe('Editor', () => {
  let tabsStore, shallow, mount;

  afterEach(() => {
    mount.cleanUp();
  });

  beforeEach(() => {
    shallow = createShallow({
      dive: true
    });
    mount = createMount();
    tabsStore = new TabsStore();
    tabsStore.addTab();
    tabsStore.tabs[0].body = "";
  });

  it('matches default state snapshot', () => {
    const wrapper = shallow(<Editor tab={tabsStore.tabs[0]} tabsStore={tabsStore} />);
    expect(wrapper).toMatchSnapshot();
  });

  xit('should not add preview to toolbar by default', () => {
    const wrapper = shallow(<Editor tab={tabsStore.tabs[0]} tabsStore={tabsStore} />);
    let props = wrapper.find('SimpleMDEEditor').props();
    expect(props).toEqual();
  });

});