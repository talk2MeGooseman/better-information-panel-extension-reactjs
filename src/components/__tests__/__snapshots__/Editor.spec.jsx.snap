// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editor matches default state snapshot 1`] = `
<div
  className="sc-bdVaJa lcwGZi"
>
  <SimpleMDEEditor
    onChange={[Function]}
    options={
      Object {
        "autoDownloadFontAwesome": false,
        "insertTexts": Object {
          "toggleUnorderedList": Array [
            "* ",
            "",
          ],
        },
        "previewRender": [Function],
        "spellChecker": true,
        "status": false,
        "toolbar": Array [
          "heading-1",
          "heading-2",
          "heading-3",
          "bold",
          "italic",
          "strikethrough",
          "code",
          "quote",
          "unordered-list",
          "ordered-list",
          "image",
          "table",
          "horizontal-rule",
          Object {
            "action": [Function],
            "className": "fa fa-check",
            "name": "check-list",
            "title": "Check List",
          },
        ],
      }
    }
    value=""
  />
</div>
`;
